frontend/src/actions.js

```diff
@@ -17,4 +17,111 @@ const logout = () =>
+  const arr = JSON.parse(articles);
+
+  const obj = {};
+  const uniqueCounter = {};
+
+  for (let i = 0; i < arr.length; i += 1) {
+    const c = arr[i]; // current element
+    if (c.title) {
+      const path = `${c.topic}|${c.sub_topic || DEFAULT_DISPLAY_SUBTOPIC}|${
+        c.title
+      }`;
+      const key = slug(path);
+      let finalPath = key;
+      if (!obj[key]) {
+        uniqueCounter[key] = 1;
+      } else {
+        finalPath += `-${uniqueCounter[key]}`;
+        uniqueCounter[key] += 1;
+      }
+      obj[finalPath] = c._id;
+    }
+  }
+
+  return obj;
+};
+
+const buildArticleNumbers = articles => {
+  // with this table i can quickly find element in allArticles if i know id
+  const arr = JSON.parse(articles);
+  const obj = {};
+  for (let i = 0; i < arr.length; i += 1) {
+    const c = arr[i]; // current element
+    obj[c._id] = i;
+  }
+
+  return obj;
+};
+
+const buildTopicTree = articles => {
+  const arr = JSON.parse(articles);
+  const tree = {};
+  const uniqueCounter = {};
+  for (let i = 0; i < arr.length; i += 1) {
+    const c = arr[i];
+    const slugTopic = slug(c.topic);
+    if (!tree[slugTopic]) {
+      tree[slugTopic] = { titleDisplay: c.topic };
+    }
+    const slugSubtopic = slug(c.sub_topic || DEFAULT_DISPLAY_SUBTOPIC);
+    if (!tree[slugTopic][slugSubtopic]) {
+      tree[slugTopic][slugSubtopic] = {
+        titleDisplay: c.sub_topic || DEFAULT_DISPLAY_SUBTOPIC,
+      };
+    }
+    const slugTitle = slug(c.title);
+    const uniqueKey = `${slugTopic}|${slugSubtopic}|${slugTitle}`;
+    if (!tree[slugTopic][slugSubtopic][slugTitle]) {
+      tree[slugTopic][slugSubtopic][slugTitle] = { titleDisplay: c.title };
+      uniqueCounter[`${slugTopic}|${slugSubtopic}|${slugTitle}`] = 1;
+    } else {
+      tree[slugTopic][slugSubtopic][
+        `${slugTitle}-${uniqueCounter[uniqueKey]}`
+      ] = { titleDisplay: c.title };
+      uniqueCounter[uniqueKey] += 1;
+    }
+  }
+
+  return tree;
+};
+
+function checkLocalStorage() {
+  // production version :
+  // if (!localStorage.getItem('allArticles'))
+
+  // development version:
+  if (true) {
+    console.log('request articles from backend');
+    console.time('load-cms');
+    fetch('/api/v1/articles/')
+      .then(res => res.json())
+      .then(data => {
+        localStorage.setItem('allArticles', JSON.stringify(data));
+        console.log('articles data received');
+        console.timeEnd('load-cms');
+        const articles = localStorage.getItem('allArticles');
+        const pathTable = buildPathTable(articles);
+        localStorage.setItem('pathTable', JSON.stringify(pathTable));
+        const articleIndex = buildArticleNumbers(articles);
+        localStorage.setItem('articleIndex', JSON.stringify(articleIndex));
+        const tree = buildTopicTree(articles);
+        localStorage.setItem('tree', JSON.stringify(tree));
+        // this == <App />
+        this.setState({ cmsReady: true });
```

Comment:
noooo ... only just spotted this little doozy. horrible, no. :grin: 
Return a promise and resolve true when load finishes - or false if fails.

Reply:
why horrible? is not allowed to use `this.setState` ?  promise cool but one note; data load runs once when i first visit the site. (local storage empty).  A hour or a week later, 2nd or 1000th time i visit site, local storage is not empty, data load does not run but to render cms i still need to check cms ready or not?

Reply:
Functions should be as Independent as possible and `actions` has no business in state management.